### PR TITLE
add `trim_split_whitespace`

### LINKED ISSUE #8521: New lint: calling `trim()` then `split_whitespace()`
### What it does

Warns about calling `str::trim` (and variants) before `str::split_whitespace` (and variants). `split_whitespace` already ignores leading and trailing whitespace.

These two lines produce the same output:

```rust
    println!("{:?}", " A B C ".trim().split_whitespace().collect::<Vec<_>>());
    println!("{:?}", " A B C ".split_whitespace().collect::<Vec<_>>());
```
which is
```
["A", "B", "C"]
["A", "B", "C"]
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=81816ae060a7540e35cc21a043ac63ed)

### Lint Name

trim_split_whitespace

### Category

pedantic

### Advantage

- The `trim` call is needless
- The developer might not realize that `split_whitespace` works this way

### Drawbacks

_No response_

### Example

```rust
" A B C ".trim().split_whitespace()
```

Could be written as:

```rust
" A B C ".split_whitespace()
```

### PR DESCRIPTION
**Pull request by FoseFx**
Closes #8521

changelog: [`trim_split_whitespace`]

### CODE REVIEW

File: clippy_lints/src/strings.rs [hunk 1]
```diff
@@ -451,3 +451,49 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+
+impl<'tcx> LateLintPass<'tcx> for TrimSplitWhitespace {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'_>) {
+        if_chain! {
+            if let ExprKind::MethodCall(path, [recv], split_ws_span) = expr.kind;
+            if path.ident.name == sym!(split_whitespace);
+            if let ExprKind::MethodCall(path, [recv], trim_span) = recv.kind;
+            if let trim_fn_name @ ("trim" | "trim_start" | "trim_end") = path.ident.name.as_str();
+            let recv_ty = cx.typeck_results().expr_ty(recv).peel_refs();
+            if recv_ty.is_str() || is_string(cx, recv);
```
Comment (FoseFx) [inline clippy_lints/src/strings.rs:485]: This only checks for `&str` and `String`. So anything implementing `Deref<Target=str>` will not be linted. Do we want that?

Comment (mikerite) [reply clippy_lints/src/strings.rs:485]: Yes. It should run if those methods are chained regardless of whether it's through a deref.

Comment (FoseFx) [reply clippy_lints/src/strings.rs:485]: I don't think this is what clippy usually does. [In this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3b103af0ca46aaf5556113e9a8a7a402) only the `&str` version is triggering a `str_to_string` lint. The version using a wrapping struct implementing `Deref<Target=str>` does not.

Comment (mikerite) [reply clippy_lints/src/strings.rs:485]: The `str_to_string` lint only warns on `&str`s because it is specially about calling `to_string` on `str` -  [str_to_string docs](https://rust-lang.github.io/rust-clippy/master/index.html#str_to_string) 

This lint is applicable to any type that `.trim().split_whitespace()` can be used on. You just need to be careful with your suggestion if the type has it's own `split_whitespace` method.

File: clippy_lints/src/strings.rs [hunk 2]
```diff
@@ -451,3 +451,65 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+        if let Some(deref_trait_id) = cx.tcx.get_diagnostic_item(sym::Deref);
+        if implements_trait(cx, ty, deref_trait_id, &[]);
+        if let Some(target) = get_associated_type(cx, ty, deref_trait_id, "Target");
+        if target.is_str();
+        then {
+            true
+        }
+        else {
+            false
+        }
```
Comment (flip1995) [inline clippy_lints/src/strings.rs:513]:
```suggestion
        then {
            target.is_str()
        }
        else {
            false
        }
```

File: clippy_lints/src/strings.rs [hunk 3]
```diff
@@ -451,3 +451,65 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+    /// ```rust
+    /// " A B C ".trim().split_whitespace();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// " A B C ".split_whitespace();
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub TRIM_SPLIT_WHITESPACE,
+    pedantic,
```
Comment (flip1995) [inline clippy_lints/src/strings.rs:472]: This could be a `style` lint IMO.

File: clippy_lints/src/strings.rs [hunk 4]
```diff
@@ -451,3 +451,65 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+    ///
+    /// ### Example
+    /// ```rust
+    /// " A B C ".trim().split_whitespace();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// " A B C ".split_whitespace();
+    /// ```
+    #[clippy::version = "1.61.0"]
```
Comment (flip1995) [inline clippy_lints/src/strings.rs:470]:
```suggestion
    #[clippy::version = "1.62.0"]
```


This will only get into the 1.62 nightly.

File: clippy_lints/src/strings.rs [hunk 5]
```diff
@@ -451,3 +451,65 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+    pub TRIM_SPLIT_WHITESPACE,
+    pedantic,
+    "using `str::trim()` or alike before `str::split_whitespace`"
+}
+declare_lint_pass!(TrimSplitWhitespace => [TRIM_SPLIT_WHITESPACE]);
+
+impl<'tcx> LateLintPass<'tcx> for TrimSplitWhitespace {
```
Comment (flip1995) [inline clippy_lints/src/strings.rs:477]: Could this lint be implemented in the `methods` module? I think this would greatly simplify the lint implementation.

File: tests/ui/trim_split_whitespace.rs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    // Custom
+    let _ = Custom().trim().split_whitespace(); // should not trigger lint
+
+    // Deref<Target=str>
+    let s = DerefStr(" A B C ");
+    let _ = s.trim().split_whitespace(); // should trigger lint
+
+    // Deref<Target=str> + custom impl
+    let s = DerefStrAndCustom(" A B C ");
+    let _ = s.trim().split_whitespace(); // should not trigger lint
```
[the remark this answers is not in the document]
Comment (flip1995) [reply tests/ui/trim_split_whitespace.rs:54]: I think the only clean way to do this is to add `rustc_diagnostic_item` attributes to those functions in `rust-lang/rust`. It's pretty easy to do, see https://github.com/rust-lang/rust/pull/87170

[the remark this answers is not in the document]
Comment (FoseFx) [reply tests/ui/trim_split_whitespace.rs:54]: Alright I get to work, should I `r?` you on the PR?

**Comment by flip1995** [comment]
To use the diagnostic items in this PR, you can try to bump the nightly version in the `rust-toolchain` file to tomorrow 2022-03-26. I think that should work without breaking anything else in Clippy.

File: clippy_lints/src/strings.rs [hunk 7]
```diff
@@ -451,3 +451,64 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+                    "split_whitespace()".to_string(),
+                    Applicability::MachineApplicable,
+                );
+            }
+        }
+    }
+}
+
+/// does ty implement Deref<Target=str>?
+fn implements_deref_str<'t>(cx: &LateContext<'t>, ty: Ty<'t>) -> bool {
```
Comment (flip1995) [inline clippy_lints/src/strings.rs:502]: With diag items, not sure if this is even necessary, because you only check for the functions implemented for `str`.

**Comment by FoseFx** [comment]
There is a different test failing on `nightly-2022-03-26`. I guess I wait until that is sorted out on master?

**Comment by mikerite** [comment]
Looks like the tests are missing the edge case where the type implements `Deref<Target=str>` and has it's own `split_whitespace` but does not have it's own `trim`.

Otherwise it's looking good. I tested on a bunch of top crates. I found 3 warnings in `procfs` and fix worked.

```
---> procfs-0.12.0/src/net.rs:570:27                                  
warning: found call to `str::trim` before `str::split_whitespace`     
   --> src/net.rs:570:27
    |
570 |         let mut split = s.trim().split_whitespace();
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `trim()`: `split_whitespace()`
    |
    = note: requested on the command line with `-W clippy::trim-split-whitespace`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#trim_split_whitespace


---> procfs-0.12.0/src/diskstats.rs:103:26                            
warning: found call to `str::trim` before `str::split_whitespace`     
   --> src/diskstats.rs:103:26
    |
103 |         let mut s = line.trim().split_whitespace();
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `trim()`: `split_whitespace()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#trim_split_whitespace


---> procfs-0.12.0/src/locks.rs:146:26                                
warning: found call to `str::trim` before `str::split_whitespace`     
   --> src/locks.rs:146:26
    |
146 |         let mut s = line.trim().split_whitespace();
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `trim()`: `split_whitespace()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#trim_split_whitespace
```

**Comment by FoseFx** [comment]
*poke*

**Comment by FoseFx** [comment]
*poke*

File: clippy_lints/src/strings.rs [hunk 8]
```diff
@@ -451,3 +452,58 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+            if let trim_fn_name @ ("trim" | "trim_start" | "trim_end") = path.ident.name.as_str();
+            if let Some(trim_def_id) = tyckres.type_dependent_def_id(split_recv.hir_id);
+            if is_one_of_trim_diagnostic_items(cx, trim_def_id);
+            then {
+                span_lint_and_sugg(
+                    cx,
+                    TRIM_SPLIT_WHITESPACE,
+                    split_ws_span.with_lo(trim_span.lo()),
+                    format!("found call to `str::{}` before `str::split_whitespace`", trim_fn_name).as_str(),
+                    format!("remove `{}()`", trim_fn_name).as_str(),
```
Comment (flip1995) [inline clippy_lints/src/strings.rs:496]:
```suggestion
                    &format!("found call to `str::{}` before `str::split_whitespace`", trim_fn_name),
                    &format!("remove `{}()`", trim_fn_name),
```

File: clippy_lints/src/strings.rs [hunk 9]
```diff
@@ -451,3 +452,58 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
...
+            if cx.tcx.is_diagnostic_item(sym::str_split_whitespace, split_ws_def_id);
+            if let ExprKind::MethodCall(path, [_trim_recv], trim_span) = split_recv.kind;
+            if let trim_fn_name @ ("trim" | "trim_start" | "trim_end") = path.ident.name.as_str();
+            if let Some(trim_def_id) = tyckres.type_dependent_def_id(split_recv.hir_id);
+            if is_one_of_trim_diagnostic_items(cx, trim_def_id);
+            then {
+                span_lint_and_sugg(
+                    cx,
+                    TRIM_SPLIT_WHITESPACE,
+                    split_ws_span.with_lo(trim_span.lo()),
```
Comment (flip1995) [inline clippy_lints/src/strings.rs:494]: You can just use the span of the `trim` function here (up until the `.` for the next method call or with `lo` of the `.` of the `trim` method call) and then use `String::new()` for the suggestion. That way you won't suggest removing `split_whitespace` to immediately add it back.

File: tests/ui/trim_split_whitespace.fixed [hunk 10]
```diff
@@ -0,0 +1,91 @@
+// run-rustfix
+#![warn(clippy::trim_split_whitespace)]
+#![allow(clippy::let_unit_value)]
+
+struct Custom();
```
Comment (flip1995) [inline tests/ui/trim_split_whitespace.fixed:5]:
```suggestion
struct Custom;
```

File: tests/ui/trim_split_whitespace.fixed [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    let _ = " A B C ".split_whitespace(); // should trigger lint
+    let _ = " A B C ".split_whitespace(); // should trigger lint
+
+    // String
+    let _ = (" A B C ").to_string().split_whitespace(); // should trigger lint
+    let _ = (" A B C ").to_string().split_whitespace(); // should trigger lint
+    let _ = (" A B C ").to_string().split_whitespace(); // should trigger lint
+
+    // Custom
+    let _ = Custom().trim().split_whitespace(); // should not trigger lint
```
Comment (flip1995) [inline tests/ui/trim_split_whitespace.fixed:72]:
```suggestion
    let _ = Custom.trim().split_whitespace(); // should not trigger lint
```

**Review by flip1995 — COMMENTED** [review]
LGTM me overall. Small suggestion on how to improve the suggestion.

**Comment by FoseFx** [comment]
:+1:

**Review by flip1995 — APPROVED** [review]
Thanks! LGTM.

Can you squash the commits? After that, this is ready to get merged.

**Comment by FoseFx** [comment]
Done 👍

**Comment by flip1995** [comment]
Thanks! Sorry reviews take so long currently... 

@bors r+
